src/esphomelib/sensor/tx20_component.cpp

```diff
@@ -0,0 +1,160 @@
+// Based on:
+
+#include "esphomelib/defines.h"
+
+#ifdef USE_TX20
+
+#include "esphomelib/espmath.h"
+#include "esphomelib/log.h"
+#include "esphomelib/text_sensor/text_sensor.h"
+#include "tx20_component.h"
+
+#ifdef ARDUINO_ARCH_ESP32
+  #define GPIO_STATUS_W1TC_ADDRESS 0x24
+#endif
+
+ESPHOMELIB_NAMESPACE_BEGIN
+
+namespace sensor {
+
+static const char *TAG = "sensor.tx20";
+
+const char UNIT_KMH[] = "km/h";
+const char ICON_WIND_SPEED[] = "mdi:weather-windy";
+const char ICON_WIND_DIRECTION[] = "mdi:sign-direction";
+const char EMPTY[] = "";
+const uint16_t TX20_BIT_TIME = 1220;   // microseconds
+const uint8_t TX20_RESET_VALUES = 60;  // seconds
+
+const std::vector<std::string> DIRECTIONS = {"N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
+                                             "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"};
+
+static bool tx20_available = false;
+static float tx20_wind_speed_kmh = 0;
+static uint8_t tx20_wind_direction = 0;
+static bool firstInstance = false;
+
+TX20Component::TX20Component(const std::string &wind_speed_name, const std::string &wind_direction_name,
+                             const std::string &wind_direction_text_name, GPIOPin *pin, uint32_t update_interval)
+    : PollingComponent(update_interval),
+      wind_speed_sensor_(new TX20WindSpeedSensor(wind_speed_name, this)),
+      wind_direction_sensor_(new TX20WindDirectionSensor(wind_direction_name, this)),
+      wind_direction_text_sensor_(new TX20WindDirectionTextSensor(wind_direction_text_name)),
+      pin_(pin) {
+  ESP_LOGV(TAG, "Start TX20Component...");
+  if(firstInstance){
+    ESP_LOGE(TAG, "Multiple TX20 Sensors are not supported!!");
+  }
+  firstInstance = true;
+}
+
+void TX20Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up TX20...");
+  this->pin_->setup();
+
+  attachInterrupt(this->pin_->get_pin(), TX20Component::start_read_, RISING);
+  global_tx20_ = this;
+}
+void TX20Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "TX20:");
+  LOG_PIN(" Pin:", TX20Component::pin_);
+  LOG_UPDATE_INTERVAL(this);
+}
+float TX20Component::get_setup_priority() const {
+  return setup_priority::HARDWARE_LATE;
+}
+
+void TX20Component::update() {
+  if (tx20_available) {
+    ESP_LOGV(TAG, "Updating TX20...");
+    std::string direction = DIRECTIONS.at(tx20_wind_direction);
```

Comment:
Check if `DIRECTIONS` contains the key first. We do not want to crash the system if we get a faulty reading.